File: 0082. Remove Duplicates from Sorted List II/Remove Duplicates from Forted List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        pre, cur = dummy, head
        
        while cur:
            while cur.next and cur.val == cur.next.val:
                cur = cur.next
            if pre.next == cur:
                pre = cur
            else:
                pre.next = cur.next
            cur = cur.next
        
        return dummy.next

File: 0082. Remove Duplicates from Sorted List II/Remove Duplicates from Forted List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode *dummy = new ListNode(0);
        dummy->next = head;
        ListNode *pre = dummy, *cur = head;
        
        while (cur) {
            while (cur->next && cur->val == cur->next->val) cur = cur->next;
            if (pre->next == cur) pre = cur;
            else pre->next = cur->next;
            cur = cur->next;
        }
        
        return dummy->next;
    }
};